File: problems-extend-adapt/3801-3900/3833_counting-indices-that-outscore-their-suffix/solution.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var countOutscoringIndices = function (nums) {
    // Suffix sums stay within 99 * 100 = 9900 and cross-products within
    // 100 * 99 = 9900, so plain numbers hold every integer here exactly,
    // far inside the 2^53 safe range.
    const n = nums.length;
    let count = 0;
    // Walk backward from the second-to-last index, carrying the sum of
    // the strict suffix after i; the comparison nums[i] > sum / (n - 1 - i)
    // is exactly nums[i] * (n - 1 - i) > sum in integer arithmetic.
    let suffix = 0;
    for (let i = n - 2; i >= 0; i--) {
        suffix += nums[i + 1];
        if (nums[i] * (n - 1 - i) > suffix) {
            count += 1;
        }
    }
    return count;
};

File: problems-extend-adapt/3801-3900/3833_counting-indices-that-outscore-their-suffix/solution.rs
impl Solution {
    // Suffix sums stay within 99 * 100 = 9900 and cross-products within
    // 100 * 99 = 9900, so i32 arithmetic carries both without overflow.
    pub fn count_outscoring_indices(nums: Vec<i32>) -> i32 {
        let n = nums.len();
        let mut count = 0i32;
        // Walk backward from the second-to-last index, carrying the sum of
        // the strict suffix after i; the comparison nums[i] > sum / (n - 1 - i)
        // is exactly nums[i] * (n - 1 - i) > sum in integer arithmetic.
        let mut suffix = 0i32;
        for i in (0..n - 1).rev() {
            suffix += nums[i + 1];
            if nums[i] * (n as i32 - 1 - i as i32) > suffix {
                count += 1;
            }
        }
        count
    }
}
